File: 6. Secondary Development/Arduino Developmemt/Sensor-extension Game/Program Files/9. Sound Detection and Placement/LightSensor_put/ESPMax.cpp
#include "ESPMax.h"
#include "_espmax.h"
#include "LobotSerialServoControl.h"

#define SERVO_SERIAL_RX   35
#define SERVO_SERIAL_TX   12
#define receiveEnablePin  13
#define transmitEnablePin 14
HardwareSerial HardwareSerial(2);
LobotSerialServoControl BusServo(HardwareSerial,receiveEnablePin,transmitEnablePin);

float ORIGIN[3] ={ 0, -(L1 + L3 + L4), (L0 + L2)};
float positions[3];

void ESPMax_init(){
    BusServo.OnInit();
    HardwareSerial.begin(115200,SERIAL_8N1,SERVO_SERIAL_RX,SERVO_SERIAL_TX);
}

int set_servo_in_range(int servo_id, int p, int duration){
    if(servo_id == 3 & p < 470) p = 470;
    if(servo_id == 2 & p > 700) p = 700;
    BusServo.LobotSerialServoMove(servo_id, p, duration);
    return int(1);
}

float* position_to_pulses(float pos[3], float* pul){
    float angles[3];
    inverse(pos,angles);
    deg_to_pulse(angles,pul);
    return pul;
}

float* pulses_to_position(float pul[3], float* pos){
    float joints[3];
    pulse_to_deg(pul,joints);
    forward(joints,pos);
    return pos;
}

int set_position(float pos[3], int duration){
    float x = pos[0];
    float y = pos[1];
    float z = pos[2];
    if(z > 255) z = 255;
    if(sqrt(x*x + y*y) < 50) return int(0);
    float angles[3];
    inverse(pos,angles);
    float pul[3];
    deg_to_pulse(angles,pul);
    for(int i=0; i<3; i++){
        positions[i] = pul[i];
        BusServo.LobotSerialServoMove(i+1,pul[i],duration);
        delay(2);
    }
    return int(1);
}

void set_position_with_speed(float pos[3], int speeds){
    float sum = 0.0;
    for(int i=0; i<3; i++){
        sum += (pos[i]-positions[i])*(pos[i]-positions[i]);
    }
    float distance = sqrt(sum);
    int duration = int(distance / speeds);
    set_position(pos, duration);
}

int set_position_relatively(float values[3], int duration){
    float x = positions[0];
    float y = positions[1];
    float z = positions[2];
    float dx = values[0];
    float dy = values[1];
    float dz = values[2];
    x += dx;
    y += dy;
    z += dz;
    float pos[3]={x,y,z};
    return set_position(pos, duration);  
}

void go_home(int duration){
    set_position(ORIGIN, duration);
}

void teaching_mode(){
    for(int i=1; i<4; i++){
        BusServo.LobotSerialServoUnload(i);
    }
}

float* read_position(float* pos){
    float pul[3];
    for(int i=0; i<3; i++){
        pul[i] = BusServo.LobotSerialServoReadPosition(i+1);
    }
    pulses_to_position(pul,pos);
}

File: Appendix/9. Handle Control Programs/espmax.py
import time
import math
import __espmax
from BusServo import BusServo

L0 = 84.0
L1 = 8.2
L2 = 128.0
L3 = 138.0
L4 = 16.8


class ESPMax:
    ORIGIN = 0, -(L1 + L3 + L4), L0 + L2

    def __init__(self, bus_servo=BusServo(), origin=ORIGIN):
      self.bus_servo = bus_servo
      self.origin = origin
      self.position = self.origin
      self.joints = 120, 90, 0
      self.servos = 500, 500, 500  # [servo id 1, servo id 2, servo id 3]
      self.last_position = self.position
      self.distance = 0
      self.duration = 0
      self.L4 = L4

    def set_servo_in_range(self, servo_id, p, duration):
        if servo_id == 3 and p < 470:
            p = 470
        if servo_id == 2 and p > 700:
            p = 700

        self.bus_servo.run(servo_id, int(p), duration)
        return True
    
    def position_to_pulses(self, position):
        angles = __espmax.inverse(self.L4, position)
        pulses = __espmax.deg_to_pulse(angles)
        return pulses
  
    def pulses_to_position(self, pulses):
        joints = __espmax.pulse_to_deg(pulses)
        x,y,z = __espmax.forward(self.L4, joints)
        return -int(x+0.5),int(y+0.5),int(z+0.5)
  
    def set_position(self, position, duration):
        duration = int(duration)
        x, y, z = position
        try:
          if z > 225:
              z = 225
          if math.sqrt(x ** 2 + y ** 2) < 50:
              return None
          angles = __espmax.inverse(self.L4, (-x, y, z))
          pulses = __espmax.deg_to_pulse(angles)
          for i in range(3):
                  ret = self.set_servo_in_range(i + 1, pulses[i], duration)
                  if not ret:
                      raise ValueError("{} Out of limit range".format(pulses[i]))
          self.servos = pulses
          self.joints = angles
          self.position = x, y, z
          return True
        except:
          return False
    
    def set_position_with_speed(self, position, speed):
            old_position = self.position
            distance = math.sqrt(sum([(position[i] - old_position[i]) ** 2 for i in range(0, 3)]))
            duration = distance / max(speed, 0.001)
            self.distance = distance
            self.last_position = self.position
            self.duration = duration
            self.set_position(position, duration)
    
    def set_position_relatively(self, values, duration):
            x, y, z = self.position
            x_v, y_v, z_v = values
            x += x_v
            y += y_v
            z += z_v
            return self.set_position((x, y, z), duration)
    
    def set_servo(self, servo_id, pulse, duration):
        if not 0 < servo_id < 4:
            raise ValueError("Invalid servo id:{}".format(servo_id))
        pulse = 0 if pulse < 0 else pulse
        pulse = 1000 if pulse > 1000 else pulse
        servos = list(self.servos)
        servos[servo_id - 1] = pulse
        joints = __espmax.pulse_to_deg(servos)
        new_position = __espmax.forward(self.L4, joints)
        self.set_position(new_position, duration)

    def set_servo_with_speed(self, servo_id, pulse, speed):
        if not 0 < servo_id < 4:
            raise ValueError("Invalid servo id:{}".format(servo_id))
        pulse = 0 if pulse < 0 else pulse
        pulse = 1000 if pulse > 1000 else pulse
        servos = list(self.servos)
        servos[servo_id - 1] = pulse
        joints = __espmax.pulse_to_deg(servos)
        new_position = __espmax.forward(self.L4, joints)
        self.set_position_with_speed(new_position, speed)

    def set_servo_relatively(self, servo_id, value, duration):
        if not 0 < servo_id < 4:
            raise ValueError("Invalid servo id:{}".format(servo_id))
        index = servo_id - 1
        pulse = self.servos[index]
        pulse += value
        return self.set_servo(servo_id, pulse, duration)

    def set_joint(self, joint_id, angle, duration):
        if not 0 < joint_id < 4:
            raise ValueError("Invalid joint id:{}".format(joint_id))
        angles = list(self.joints)
        angles[joint_id - 1] = angle
        servos = __espmax.deg_to_pulse(angles)
        return self.set_servo(joint_id, servos[joint_id - 1], duration)

    def set_joint_relatively(self, joint_id, value, duration):
        if not 0 < joint_id < 4:
            raise ValueError("Invalid joint id:{}".format(joint_id))
        angles = list(self.joints)
        angles[joint_id - 1] += value
        servos = __espmax.deg_to_pulse(angles)
        return self.set_servo(joint_id, servos[joint_id - 1], duration)

    def go_home(self, duration=2000):
        self.set_position(self.origin, duration)
  
    def teaching_mode(self):
      for i in range(3):
        self.bus_servo.unload(i+1)
  
    def read_position(self):
      pulses_list = []
      for i in range(3):
        pulses = self.bus_servo.get_position(i+1)
        if pulses:
          pulses_list.append(pulses)
        else:
          n = 0
          for s in range(3):
            pulses = self.bus_servo.get_position(i+1)
            if pulses:
              pulses_list.append(pulses)
              break
            else:
              n += 1
              if n == 3:
                return False
            time.sleep_ms(5)
        time.sleep_ms(5)
      x,y,z = ESPMax.pulses_to_position(self,pulses_list)
      return x,y,z
    
    def verify_position(self, x,y,z):
      try:
        angles = __espmax.inverse(self.L4, (x,y,z))
        pulses = __espmax.deg_to_pulse(angles)
        return True
      except:
        return False















